jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassifier.scala

```diff
@@ -287,46 +284,54 @@ class XGBoostClassificationModel private[ml](
     val bBooster = dataset.sparkSession.sparkContext.broadcast(_booster)
     val appName = dataset.sparkSession.sparkContext.appName
 
-    val inputRDD = dataset.asInstanceOf[Dataset[Row]].rdd
-    val predictionRDD = dataset.asInstanceOf[Dataset[Row]].rdd.mapPartitions { rowIterator =>
-      if (rowIterator.hasNext) {
-        val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
-        Rabit.init(rabitEnv.asJava)
-        val featuresIterator = rowIterator.map(row => row.getAs[Vector](
-          $(featuresCol))).toList.iterator
-        import DataUtils._
-        val cacheInfo = {
-          if ($(useExternalMemory)) {
-            s"$appName-${TaskContext.get().stageId()}-dtest_cache-${TaskContext.getPartitionId()}"
-          } else {
-            null
+    val resultRDD = dataset.asInstanceOf[Dataset[Row]].rdd.mapPartitions { rowIterator =>
+      new AbstractIterator[Row] {
+        private var batchCnt = 0
+
+        private val batchIterImpl = rowIterator.grouped(
+          XGBoostClassificationModel.PREDICTION_BATCH_SIZE).flatMap { batchRow =>
+          if (batchCnt == 0) {
+            val rabitEnv = Array("DMLC_TASK_ID" -> TaskContext.getPartitionId().toString).toMap
+            Rabit.init(rabitEnv.asJava)
+          }
+
+          val features = batchRow.map(row => row.getAs[Vector]($(featuresCol)))
+
+          import DataUtils._
+          val cacheInfo = {
+            if ($(useExternalMemory)) {
+              s"$appName-${TaskContext.get().stageId()}-dtest_cache-" +
+                  s"${TaskContext.getPartitionId()}-batch-$batchCnt"
+            } else {
+              null
+            }
+          }
+
+          val dm = new DMatrix(
+            XGBoost.processMissingValues(features.map(_.asXGB).iterator, $(missing)),
+            cacheInfo)
+          try {
+            val Array(rawPredictionItr, probabilityItr, predLeafItr, predContribItr) =
+              producePredictionItrs(bBooster, dm)
+            produceResultIterator(batchRow.iterator,
+              rawPredictionItr, probabilityItr, predLeafItr, predContribItr)
+          } finally {
+            batchCnt += 1
+            dm.delete()
           }
         }
-        val dm = new DMatrix(
-          XGBoost.processMissingValues(featuresIterator.map(_.asXGB), $(missing)),
-          cacheInfo)
-        try {
-          val Array(rawPredictionItr, probabilityItr, predLeafItr, predContribItr) =
-            producePredictionItrs(bBooster, dm)
-          Rabit.shutdown()
-          Iterator(rawPredictionItr, probabilityItr, predLeafItr,
-            predContribItr)
-        } finally {
-          dm.delete()
+
+        override def hasNext: Boolean = batchIterImpl.hasNext
+
+        override def next(): Row = {
+          val ret = batchIterImpl.next()
```

Comment:
ah, I see, the key point is here, IIUC, with this implementation you don't need to locate prediction results in memory for a complete partition,

Reply:
Yes. And, I've tried to mock your solution with the below script. 

```scala
object MiniBatchDebug {
  import scala.collection.Iterator
  import scala.collection.mutable.ListBuffer

  private val batchSize = 3
  private val dataSize = 9

  def f(iter: Iterator[Array[Int]]): Iterator[(Array[Int], Int)] = {
    var batchCnt = 0
    val results = new ListBuffer[Iterator[(Array[Int], Int)]]
    while (iter.hasNext) {
      println(s"fetch batch num: $batchCnt")
      val (x, y) = iter.take(batchSize).duplicate
      // mock the reading behavior of batch compute
      val ret = y.toArray.map(_.sum).iterator
      results += x.zip(ret)
      batchCnt += 1
    }
    results.reduce((i1, i2) => i1 ++ i2)
  }

  def main(args: Array[String]): Unit = {
    val dataSource = Seq.tabulate(dataSize)(x => Array(x, x * 2))
    val sourceIter: Iterator[Array[Int]] = dataSource.iterator
    val result = f(sourceIter).map { case (x, y) => x.sum + y }
    (1 to dataSize).foreach { i =>
      println(s"fetch element $i")
      result.next()
    }
  }
}
```

Here is the print info, which demonstrated it will run all batches before real data fetching :
>  fetch batch num: 0
fetch batch num: 1
fetch batch num: 2
fetch element 1
fetch element 2
fetch element 3
fetch element 4
fetch element 5
fetch element 6
fetch element 7
fetch element 8
fetch element 9